// uploadthing/dist/fastify.mjs
import {
  toWebRequest
} from "./chunk-EMWTWK2B.mjs";
import {
  UPLOADTHING_VERSION,
  buildPermissionsInfoHandler,
  buildRequestHandler,
  createBuilder,
  formatError,
  incompatibleNodeGuard,
  initLogger
} from "./chunk-VF56OB6T.mjs";
import "./chunk-4QCB7YXI.mjs";
import "./chunk-JW5CQVQS.mjs";

// src/fastify.ts
import { getStatusCodeFromError, UploadThingError } from "@uploadthing/shared";
var createUploadthing = (opts) => createBuilder(opts);
var createRouteHandler = (fastify, opts, done) => {
  var _a;
  initLogger((_a = opts.config) == null ? void 0 : _a.logLevel);
  incompatibleNodeGuard();
  const requestHandler = buildRequestHandler(
    opts,
    "fastify"
  );
  const getBuildPerms = buildPermissionsInfoHandler(opts);
  const POST = async (req, res) => {
    const proto = req.headers["x-forwarded-proto"] ?? "http";
    const url = new URL(req.url, `${proto}://${req.headers.host}`);
    const response = await requestHandler({
      nativeRequest: toWebRequest(req, url),
      originalRequest: req,
      res,
      event: void 0
    });
    if (response instanceof UploadThingError) {
      void res.status(getStatusCodeFromError(response)).headers({
        "x-uploadthing-version": UPLOADTHING_VERSION
      }).send(formatError(response, opts.router));
      return;
    }
    if (response.status !== 200) {
      void res.status(500).headers({
        "x-uploadthing-version": UPLOADTHING_VERSION
      }).send("An unknown error occured");
      return;
    }
    void res.status(response.status).headers({
      "x-uploadthing-version": UPLOADTHING_VERSION
    }).send(response.body);
  };
  const GET = async (req, res) => {
    void res.status(200).headers({
      "x-uploadthing-version": UPLOADTHING_VERSION
    }).send(getBuildPerms());
  };
  fastify.post("/api/uploadthing", POST).get("/api/uploadthing", GET);
  done();
};
var fastifyUploadthingPlugin = createRouteHandler;
export {
  createRouteHandler,
  createUploadthing,
  fastifyUploadthingPlugin
};
