// uploadthing/dist/next-legacy.mjs
import {
  toWebRequest
} from "./chunk-EMWTWK2B.mjs";
import {
  UPLOADTHING_VERSION,
  buildPermissionsInfoHandler,
  buildRequestHandler,
  createBuilder,
  formatError,
  incompatibleNodeGuard,
  initLogger
} from "./chunk-VF56OB6T.mjs";
import "./chunk-4QCB7YXI.mjs";
import "./chunk-JW5CQVQS.mjs";

// src/next-legacy.ts
import { getStatusCodeFromError, UploadThingError } from "@uploadthing/shared";
var createUploadthing = (opts) => createBuilder(opts);
var createRouteHandler = (opts) => {
  var _a;
  initLogger((_a = opts.config) == null ? void 0 : _a.logLevel);
  incompatibleNodeGuard();
  const requestHandler = buildRequestHandler(
    opts,
    "nextjs-pages"
  );
  const getBuildPerms = buildPermissionsInfoHandler(opts);
  return async (req, res) => {
    if (req.method === "GET") {
      const perms = getBuildPerms();
      res.status(200).json(perms);
      return;
    }
    const proto = req.headers["x-forwarded-proto"] ?? "http";
    const url = new URL(req.url ?? "/", `${proto}://${req.headers.host}`);
    const response = await requestHandler({
      nativeRequest: toWebRequest(req, url),
      originalRequest: req,
      res,
      event: void 0
    });
    res.setHeader("x-uploadthing-version", UPLOADTHING_VERSION);
    if (response instanceof UploadThingError) {
      res.status(getStatusCodeFromError(response));
      res.setHeader("x-uploadthing-version", UPLOADTHING_VERSION);
      return res.json(formatError(response, opts.router));
    }
    if (response.status !== 200) {
      res.status(500);
      return res.send("An unknown error occured");
    }
    res.status(response.status);
    return res.json(response.body);
  };
};
var createNextPageApiHandler = createRouteHandler;
export {
  createNextPageApiHandler,
  createRouteHandler,
  createUploadthing
};
